hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java

```diff
@@ -201,102 +211,119 @@ default ReactiveConnection getReactiveConnection(SharedSessionContractImplemento
+								delegate().generateIdentityInsertString( getFactory().getSqlStringGenerationContext(), notNull ),
+								session
+						)
+						.thenCompose(
+								id -> loop(
+										1, span,
+										table -> insertReactive(
+												id,
+												fields,
+												notNull,
+												table,
+												delegate().generateInsertString( notNull, table ),
+												session
+										)
+								).thenApply( v -> id )
+						);
+					}
+					else {
+						// For the case of dynamic-insert="false", use the static SQL
+						return insertReactive(
+								fields,
+								delegate().getPropertyInsertability(),
+								delegate().getSQLIdentityInsertString(),
+								session
+						)
+						.thenCompose(
+								id -> loop(
+										1, span,
+										table -> insertReactive(
+												id,
+												fields,
+												delegate().getPropertyInsertability(),
+												table,
+												delegate().getSQLInsertStrings()[table],
+												session
+										)
+								).thenApply( v -> id )
+						);
+					}
+		} );
+	}
 
-		final int span = delegate().getTableSpan();
-		if ( delegate().getEntityMetamodel().isDynamicInsert() ) {
-			// For the case of dynamic-insert="true", we need to generate the INSERT SQL
-			boolean[] notNull = delegate().getPropertiesToInsert( fields );
-			return insertReactive(
-					fields,
-					notNull,
-					//this differs from core, but it's core that should be changed:
-					delegate().generateIdentityInsertString( getFactory().getSqlStringGenerationContext(), notNull ),
-					session
-			)
-			.thenCompose(
-					id -> loop(
-							1, span,
-							table -> insertReactive(
-									id,
-									fields,
-									notNull,
-									table,
-									delegate().generateInsertString( notNull, table ),
-									session
-							)
-					).thenApply( v -> id )
-			);
-		}
-		else {
-			// For the case of dynamic-insert="false", use the static SQL
-			return insertReactive(
-					fields,
-					delegate().getPropertyInsertability(),
-					delegate().getSQLIdentityInsertString(),
-					session
-			)
-			.thenCompose(
-					id -> loop(
-							1, span,
-							table -> insertReactive(
-									id,
-									fields,
-									delegate().getPropertyInsertability(),
-									table,
-									delegate().getSQLInsertStrings()[table],
-									session
-							)
-					).thenApply( v -> id )
-			);
+	default CompletionStage<Void> reactivePreInsertInMemoryValueGeneration(Object[] fields, Object object, SharedSessionContractImplementor session) {
+		CompletionStage<Void> stage = voidFuture();
+		if ( getEntityMetamodel().hasPreInsertGeneratedValues() ) {
+			final InMemoryValueGenerationStrategy[] strategies = getEntityMetamodel().getInMemoryValueGenerationStrategies();
+			for ( int i = 0; i < strategies.length; i++ ) {
+				final int index = i;
+				final InMemoryValueGenerationStrategy strategy = strategies[i];
+				if ( strategy != null && strategy.getGenerationTiming().includesInsert() ) {
```

Comment:
What happens to in-memory (nonreactive) value generators?

Reply:
Not sure I understand the question. But I've just moved the existing code